python/pyspark/sql/dataframe.py

```diff
@@ -1891,14 +1892,19 @@ def toPandas(self):
                       "if using spark.sql.execution.arrow.enable=true"
                 raise ImportError("%s\n%s" % (e.message, msg))
         else:
+            pdf = pd.DataFrame.from_records(self.collect(), columns=self.columns)
+
             dtype = {}
             for field in self.schema:
                 pandas_type = _to_corrected_pandas_type(field.dataType)
-                if pandas_type is not None:
+                # SPARK-21766: if an integer field is nullable and has null values, it can be
+                # inferred by pandas as float column. Once we convert the column with NaN back
+                # to integer type e.g., np.int16, we will hit exception.
```

Comment:
BTW, It'd be nicer if we add some comments saying like .. so, we use the inferred float type, not the corrected type from the schema.

Reply:
Added. Thanks.